### PR TITLE
Implementation of Distinct

### PR DESCRIPTION
**Pull request by CPWstatic**
GO ... YIELD DISTINCT ...

### CODE REVIEW

File: src/graph/GoExecutor.cpp [hunk 1]
```diff
@@ -75,6 +79,13 @@ void GoExecutor::execute() {
         onEmptyInputs();
         return;
     }
+    if (distinct_) {
+        std::unordered_set<VertexID> uniqID;
+        for (auto id : starts_) {
+            uniqID.emplace(id);
+        }
+        starts_ = std::vector<VertexID>(uniqID.begin(), uniqID.end());
```
Comment (whitewum) [inline src/graph/GoExecutor.cpp]: what if distinct & order by?

File: src/graph/test/GoTest.cpp [hunk 2]
```diff
@@ -154,5 +154,24 @@ TEST_F(GoTest, DISABLED_OneStepInOutBound) {
     }
 }
 
+TEST_F(GoTest, Distinct) {
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Boris Diaw"];
+        auto *fmt = "GO FROM %ld OVER like "
+                    "| GO FROM $-.id OVER like | GO FROM $-.id OVER serve "
+                    "YIELD DISTINCT serve._dst, $$[team].name";
```
Comment (whitewum) [inline src/graph/test/GoTest.cpp]: try the return is null

File: src/parser/parser.yy [hunk 3]
```diff
@@ -95,6 +95,7 @@ class GraphScanner;
 %token KW_ROLES KW_BY
 %token KW_TTL_DURATION KW_TTL_COL
 %token KW_ORDER KW_ASC
+%token KW_DISTINCT
```
Comment (whitewum) [inline src/parser/parser.yy]: better to add a case in parser ut.

File: src/graph/GoExecutor.cpp [hunk 4]
```diff
@@ -530,7 +554,15 @@ std::unique_ptr<InterimResult> GoExecutor::setupInterimResult(RpcResponse &&rpcR
                     LOG(FATAL) << "Unknown VariantType: " << column.which();
             }
         }
-        rsWriter->addRow(writer);
+        std::string encode = writer.encode();
+        if (distinct_) {
+            if (uniqResult->count(encode) == 0) {
+                uniqResult->emplace(encode);
```
Comment (whitewum) [inline src/graph/GoExecutor.cpp]: uniqResult // only emplace the first encode. right?

Comment (CPWstatic) [reply src/graph/GoExecutor.cpp]: Yes.

File: src/graph/GoExecutor.cpp (hunk 1 above)
Comment (CPWstatic) [reply src/graph/GoExecutor.cpp]: Order by is standalone statement, they will not affect each other.

File: src/parser/test/ParserTest.cpp [hunk 5]
```diff
@@ -940,4 +940,22 @@ TEST(Parser, ReentrantRecoveryFromFailure) {
...
+        std::string query = "GO FROM 1 over friend "
+                            "YIELD DISTINCT friend.name as name, friend.age as age";
+        auto result = parser.parse(query);
+        ASSERT_TRUE(result.ok()) << result.status();
+    }
+    {
+        GQLParser parser;
+        std::string query = "GO FROM 1 OVER like "
+                            "| GO FROM $-.id OVER like | GO FROM $-.id OVER serve "
+                            "YIELD DISTINCT serve._dst, $$.team.name";
```
Comment (whitewum) [inline src/parser/test/ParserTest.cpp]: syntax error: yield col1, distinct col2

Comment (CPWstatic) [reply src/parser/test/ParserTest.cpp]: That is what we want, and I add this test case in parser.

File: src/graph/GoExecutor.cpp (hunk 4 above)
Comment (dutor) [inline src/graph/GoExecutor.cpp]: You could have just called `emplace` for once, since it will tell you whether the value exists.

**Comment by whitewum** [comment]
please remember to update nGQL.md

File: src/graph/GoExecutor.cpp [hunk 6]
```diff
@@ -530,7 +554,16 @@ std::unique_ptr<InterimResult> GoExecutor::setupInterimResult(RpcResponse &&rpcR
                     LOG(FATAL) << "Unknown VariantType: " << column.which();
             }
         }
-        rsWriter->addRow(writer);
+        // TODO Consider float/double, and need to reduce mem copy.
```
Comment (whitewum) [inline src/graph/GoExecutor.cpp]: right

File: src/graph/GoExecutor.cpp [hunk 7]
```diff
@@ -530,7 +554,16 @@ std::unique_ptr<InterimResult> GoExecutor::setupInterimResult(RpcResponse &&rpcR
...
+        std::string encode = writer.encode();
+        if (distinct_) {
+            auto ret = uniqResult->emplace(encode);
+            if (ret.second) {
```
Comment (whitewum) [inline src/graph/GoExecutor.cpp]: I wonder which is more efficient?

**Pull request merged**